File: 3725.count-ways-to-choose-coprime-integers-from-rows.py
#
# @lc app=leetcode id=3725 lang=python3
#
# [3725] Count Ways to Choose Coprime Integers from Rows
#

# @lc code=start
from math import gcd
from collections import Counter
from typing import List

class Solution:
    def countCoprime(self, mat: List[List[int]]) -> int:
        MOD = 10**9 + 7
        
        # Initialize dp with the first row
        # dp[g] stores the number of ways to have a running GCD of g
        dp = [0] * 151
        
        row_counts = Counter(mat[0])
        for val, count in row_counts.items():
            dp[val] = (dp[val] + count) % MOD
            
        # Process subsequent rows
        for i in range(1, len(mat)):
            new_dp = [0] * 151
            row_counts = Counter(mat[i])
            
            # We only need to iterate through existing GCDs in dp
            # Since values are small (<=150), we can just iterate 1..150
            # Optimization: collect current valid gcds to avoid iterating 150 times if sparse
            current_gcds = [g for g in range(1, 151) if dp[g] > 0]
            
            if not current_gcds:
                return 0
                
            for g in current_gcds:
                ways = dp[g]
                for val, count in row_counts.items():
                    new_g = gcd(g, val)
                    new_dp[new_g] = (new_dp[new_g] + ways * count) % MOD
            
            dp = new_dp
            
        return dp[1]

# @lc code=end

File: 3725.count-ways-to-choose-coprime-integers-from-rows.cpp
#include <bits/stdc++.h>
using namespace std;

// @lc app=leetcode id=3725 lang=cpp
//
// [3725] Count Ways to Choose Coprime Integers from Rows
//

// @lc code=start
class Solution {
public:
    int countCoprime(vector<vector<int>>& mat) {
        static const int V = 150;
        static const int MOD = 1000000007;

        // Compute Mobius function mu[1..V] using linear sieve
        vector<int> mu(V + 1, 0), primes;
        vector<int> isComp(V + 1, 0);
        mu[1] = 1;
        for (int i = 2; i <= V; i++) {
            if (!isComp[i]) {
                primes.push_back(i);
                mu[i] = -1;
            }
            for (int p : primes) {
                long long x = 1LL * i * p;
                if (x > V) break;
                isComp[(int)x] = 1;
                if (i % p == 0) {
                    mu[(int)x] = 0;
                    break;
                } else {
                    mu[(int)x] = -mu[i];
                }
            }
        }

        // prod[d] = product over rows of count of elements divisible by d
        vector<long long> prod(V + 1, 1);

        vector<int> freq(V + 1);
        for (auto &row : mat) {
            fill(freq.begin(), freq.end(), 0);
            for (int x : row) freq[x]++;

            for (int d = 1; d <= V; d++) {
                int cnt = 0;
                for (int k = d; k <= V; k += d) cnt += freq[k];
                prod[d] = (prod[d] * cnt) % MOD;
            }
        }

        long long ans = 0;
        for (int d = 1; d <= V; d++) {
            if (mu[d] == 0) continue;
            ans = (ans + (long long)mu[d] * prod[d]) % MOD;
        }
        if (ans < 0) ans += MOD;
        return (int)ans;
    }
};
// @lc code=end
